### ISSUE TITLE
Building pikepdf on Cygwin: ‘strdup’ was not declared in this scope 

### ISSUE THREAD

**Issue by jhgarrison** [pikepdf/pikepdf#103]
While trying to build pikepdf on Cygwin (latest version, gcc 9.3.0) as part of building OCRmyPDF, I get the build failure shown below.  

      building 'pikepdf._qpdf' extension
      creating build/temp.cygwin-3.1.4-x86_64-3.7
      creating build/temp.cygwin-3.1.4-x86_64-3.7/src
      creating build/temp.cygwin-3.1.4-x86_64-3.7/src/qpdf
      gcc -Wno-unused-result -Wsign-compare -DNDEBUG -g -fwrapv -O3 -Wall -ggdb -O2 -pipe -Wall -Werror=format-security -Wp,-D_FORTIFY_SOURCE=2 -fstack-protector-strong --param=ssp-buffer-size=4 -fdebug-prefix-map=/cygdrive/d/cyg_pub/devel/python/python37/python37-3.7.7-1.x86_64/build=/usr/src/debug/python37-3.7.7-1 -fdebug-prefix-map=/cygdrive/d/cyg_pub/devel/python/python37/python37-3.7.7-1.x86_64/src/Python-3.7.7=/usr/src/debug/python37-3.7.7-1 -ggdb -O2 -pipe -Wall -Werror=format-security -Wp,-D_FORTIFY_SOURCE=2 -fstack-protector-strong --param=ssp-buffer-size=4 -fdebug-prefix-map=/cygdrive/d/cyg_pub/devel/python/python37/python37-3.7.7-1.x86_64/build=/usr/src/debug/python37-3.7.7-1 -fdebug-prefix-map=/cygdrive/d/cyg_pub/devel/python/python37/python37-3.7.7-1.x86_64/src/Python-3.7.7=/usr/src/debug/python37-3.7.7-1 -D__MISC_VISIBLE=1 -I/tmp/pip-build-env-fdjj5aaz/overlay/lib/python3.7/site-packages/pybind11/include -I/tmp/pip-build-env-fdjj5aaz/overlay/lib/python3.7/site-packages/pybind11/include -I/usr/include/python3.7m -c src/qpdf/annotation.cpp -o build/temp.cygwin-3.1.4-x86_64-3.7/src/qpdf/annotation.o -DVERSION_INFO="1.11.2" -std=c++14 -fvisibility=hidden
      In file included from src/qpdf/annotation.cpp:18:
      /tmp/pip-build-env-fdjj5aaz/overlay/lib/python3.7/site-packages/pybind11/include/pybind11/pybind11.h: In member function ‘void pybind11::cpp_function::initialize_generic(pybind11::detail::function_record*, const char*, const std::type_info* const*, pybind11::size_t)’:
      /tmp/pip-build-env-fdjj5aaz/overlay/lib/python3.7/site-packages/pybind11/include/pybind11/pybind11.h:197:21: error: ‘strdup’ was not declared in this scope; did you mean ‘strcmp’?
        197 |         rec->name = strdup(rec->name ? rec->name : "");
            |                     ^~~~~~
            |                     strcmp
      In file included from src/qpdf/annotation.cpp:18:
      /tmp/pip-build-env-fdjj5aaz/overlay/lib/python3.7/site-packages/pybind11/include/pybind11/pybind11.h: In member function ‘pybind11::class_<type_, options>& pybind11::class_<type_, options>::def_property_static(const char*, const pybind11::cpp_function&, const pybind11::cpp_function&, const Extra& ...)’:
      /tmp/pip-build-env-fdjj5aaz/overlay/lib/python3.7/site-packages/pybind11/include/pybind11/pybind11.h:1284:31: error: there are no arguments to ‘strdup’ that depend on a template parameter, so a declaration of ‘strdup’ must be available [-fpermissive]
       1284 |               rec_fget->doc = strdup(rec_fget->doc);
            |                               ^~~~~~
      /tmp/pip-build-env-fdjj5aaz/overlay/lib/python3.7/site-packages/pybind11/include/pybind11/pybind11.h:1284:31: note: (if you use ‘-fpermissive’, G++ will accept your code, but allowing the use of an undeclared name is deprecated)
      /tmp/pip-build-env-fdjj5aaz/overlay/lib/python3.7/site-packages/pybind11/include/pybind11/pybind11.h:1292:33: error: there are no arguments to ‘strdup’ that depend on a template parameter, so a declaration of ‘strdup’ must be available [-fpermissive]
       1292 |                 rec_fset->doc = strdup(rec_fset->doc);
            |                                 ^~~~~~
      /tmp/pip-build-env-fdjj5aaz/overlay/lib/python3.7/site-packages/pybind11/include/pybind11/pybind11.h: In instantiation of ‘pybind11::class_<type_, options>& pybind11::class_<type_, options>::def_property_static(const char*, const pybind11::cpp_function&, const pybind11::cpp_function&, const Extra& ...) [with Extra = {pybind11::is_method, pybind11::return_value_policy}; type_ = QPDFAnnotationObjectHelper; options = {}]’:
      /tmp/pip-build-env-fdjj5aaz/overlay/lib/python3.7/site-packages/pybind11/include/pybind11/pybind11.h:1263:80:   required from ‘pybind11::class_<type_, options>& pybind11::class_<type_, options>::def_property(const char*, const pybind11::cpp_function&, const pybind11::cpp_function&, const Extra& ...) [with Extra = {pybind11::return_value_policy}; type_ = QPDFAnnotationObjectHelper; options = {}]’
      /tmp/pip-build-env-fdjj5aaz/overlay/lib/python3.7/site-packages/pybind11/include/pybind11/pybind11.h:1252:91:   required from ‘pybind11::class_<type_, options>& pybind11::class_<type_, options>::def_property(const char*, const Getter&, const Setter&, const Extra& ...) [with Getter = pybind11::cpp_function; Setter = std::nullptr_t; Extra = {pybind11::return_value_policy}; type_ = QPDFAnnotationObjectHelper; options = {}]’
      /tmp/pip-build-env-fdjj5aaz/overlay/lib/python3.7/site-packages/pybind11/include/pybind11/pybind11.h:1234:58:   required from ‘pybind11::class_<type_, options>& pybind11::class_<type_, options>::def_property_readonly(const char*, const pybind11::cpp_function&, const Extra& ...) [with Extra = {pybind11::return_value_policy}; type_ = QPDFAnnotationObjectHelper; options = {}]’
      /tmp/pip-build-env-fdjj5aaz/overlay/lib/python3.7/site-packages/pybind11/include/pybind11/pybind11.h:1228:87:   required from ‘pybind11::class_<type_, options>& pybind11::class_<type_, options>::def_property_readonly(const char*, const Getter&, const Extra& ...) [with Getter = std::basic_string<char> (QPDFAnnotationObjectHelper::*)(); Extra = {}; type_ = QPDFAnnotationObjectHelper; options = {}]’
      src/qpdf/annotation.cpp:28:82:   required from here
      /tmp/pip-build-env-fdjj5aaz/overlay/lib/python3.7/site-packages/pybind11/include/pybind11/pybind11.h:1284:37: error: ‘strdup’ was not declared in this scope; did you mean ‘strcmp’?
       1284 |               rec_fget->doc = strdup(rec_fget->doc);
            |                               ~~~~~~^~~~~~~~~~~~~~~
            |                               strcmp
      /tmp/pip-build-env-fdjj5aaz/overlay/lib/python3.7/site-packages/pybind11/include/pybind11/pybind11.h:1292:39: error: ‘strdup’ was not declared in this scope; did you mean ‘strcmp’?
       1292 |                 rec_fset->doc = strdup(rec_fset->doc);
            |                                 ~~~~~~^~~~~~~~~~~~~~~
            |                                 strcmp
      error: command 'gcc' failed with exit status 1
      ----------------------------------------
      ERROR: Failed building wheel for pikepdf
    Failed to build pikepdf
    ERROR: Could not build wheels for pikepdf which use PEP 517 and cannot be installed directly

Examining `string.h`, which is where `strdup` is defined, I find 

    #if __MISC_VISIBLE || __POSIX_VISIBLE >= 200809 || __XSI_VISIBLE >= 4
    char    *strdup (const char *) __malloc_like __result_use_check;
    #endif
    char    *_strdup_r (struct _reent *, const char *);
    #if __POSIX_VISIBLE >= 200809
    char    *strndup (const char *, size_t) __malloc_like __result_use_check;
    #endif

But even with `-D__MISC_VISIBLE` I still get the error.

**Comment by jbarlow83** [comment]
My interpretation is that the problem has to do with Cygwin's version of `strdup` and the pybind11 header `pybind11.h` and probably other versions. More precisely, it looks like Cygwin remove `strdup` entirely for security reasons and pybind11 needs to be patched to use `strndup` instead.

First thing to check is make sure you are using pybind11 2.4.3 or newer.

You could also check if someone on the pybind11 gitter has run into this already.
https://gitter.im/pybind/Lobby

And finally, report the issue with pybind11. It looks like any attempt to include pybind11.h in cygwin will break.

**Comment by jbarlow83** [comment]
I'd be happy to update install instructions with anything you find about getting pikepdf and cygwin working together. To my knowledge it has not been attempted.

**Comment by jhgarrison** [comment]
I have pybind11 2.5.0.  I captured the entire include dependency hierarchy from the preprocessor step for `src/qpdf/annotation.cpp` (with `-MM -MF filename`)  and grepped for `strdup`, resulting in this:

    jim@HOME10 /tmp/pip-install-93lhsoke/pikepdf
    $ grep -l strdup $(cat deps.txt|tr '\n' ' ')
    /tmp/pip-build-env-fdjj5aaz/overlay/lib/python3.7/site-packages/pybind11/include/pybind11/pybind11.h
    /usr/include/python3.7m/pyconfig.h
    /usr/include/string.h
    /usr/include/python3.7m/pymem.h

I haven't written C or C++ in at least 25 years, so it's not clear to me if the fact that `string.h` (where `strdup` is declared if `__MISC_VISIBLE` is defined) appears after `pybind11.h` is critical.

**Comment by jbarlow83** [comment]
If the include order were incorrect, wouldn't pybind11 or pikepdf be broken
on all platforms?

**Comment by jhgarrison** [comment]
The solution was to specify -D_GNU_SOURCE to the compiler.  I couldn't find a way to pass this from the command line so I used the debug option to manually build annotation.cpp. Is that an option that should be the default?

**Comment by jbarlow83** [comment]
The proper solution is then to provide the compiler with `-std=gnu++14` instead of `-std=c++14`. Would you mind checking if changing this at setup.py line 90 works?

**Comment by jhgarrison** [comment]
That single change does indeed make the build succeed.  I hope that doesn't break things in other environments.

Thanks!

**Comment by jhgarrison** [comment]
On 5/16/2020 5:19 PM, jbarlow83 wrote:
> Closed #103 <https://github.com/pikepdf/pikepdf/issues/103>.
> 

Thanks for the fast turnaround.

FYI, I seem to have been successful building and installing OCRmyPDF on
Cygwin.  I've posted the instructions below to the Cygwin mailing list
asking for some volunteers to test the instructions.  If you know of
parties interested in OCRmyPDF under Cygwin please pass this on and
ask that they reply to me (jhg@acm.org).

After I get some feedback I'll provide the (possibly updated)
instructions.

----------------------------------------------------------

The instructions for installing OCRmyPDF under Cygwin are:

1. Install the following Cygwin packages:

        python36 (or later)
        python3?-devel
        python3?-pip
        python3?-lxml

     (where 3? means match the version of python3 you installed)

        gcc-g++
        ghostscript
        libexempi3
        libexempi-devel
        libffi6
        libffi-devel
        pngquant
        qpdf
        libqpdf-devel
        tesseract-ocr
        tesseract-ocr-devel

2. In a terminal, run the following commands

        pip3 install wheel
        pip3 install ocrmypdf

    Note: You may get a warning about the version of pip that gets
    installed with Cygwin being out of date.  It is not required, but
    if you want you can update pip to the latest version with

        pip3 install --upgrade pip

    But note that if you do this the command name will now be just
    'pip' instead of 'pip3'.

There is one optional dependency, "unpaper" that is currently not
available under Cygwin. Without it, certain options such as --clean
will produce an error message. However, the OCR-to-text-layer
functionality is available.  I'll take a look at building a Cygwin
version of unpaper.

-- 
Jim Garrison jhg@acm.org

**Comment by jrast** [comment]
I tried to install ocrmypdf under cygwin today and had similar issues to get pikepdf installed. To get it working I had to add these two lines directly after the `ext_modules` definition:

````python
ext_modules[0].extra_compile_args.pop(ext_modules[0].extra_compile_args.index('-std=c++14'))
ext_modules[0].extra_compile_args.append('-std=gnu++14')
````
